Add efficiencies and scale factors for embedded samples.

mburkart opened the pull request:
Add readout for efficiencies measured in embedded samples in 2016, 2017 and 2018 using analytic fits.

The efficiencies for the embedded samples are provided in separate input files than the ones for the simulated samples, since the data efficiencies corresponding to the respective measurement are used in the scale factor determination.

Changes to the code include:
 * An updated python interface. An additional flag is introduced to read out the embedded efficiencies instead of the simulated efficiencies. The interface is still backwards compatible.
 * An updated C++ interface. The updated interface is not backwards compatible since the file name is not given in the constructor anymore but the switch between embedded and simulated efficiencies is performed with a boolean flag as in the case of the python interface.
* Simple tests for the C++ and python interfaces for the embedded samples.

The measurement is documented in https://indico.cern.ch/event/864131/contributions/3644102/attachments/1946592/3229746/TauTriggerEfficiencies_Embedded.pdf. Updated comparison plots for the etau cross trigger can be found here: http://ekpwww.etp.kit.edu/~mburkart/plots_archive/2019_12_10/trigger_efficiencies/mva_comparison/index.html

Commit:
Add reading out trigger fits for embedded samples

Solved conflicts:
	interface/TauTriggerSFs2017.h
	src/TauTriggerSFs2017.cc
Conflicts were introduced because run2_SFs branch was updated to fix
warnings during compilation (commit https://github.com/cms-tau-pog/TauTriggerSFs/commit/7fc95e94ae5e98c76d21824eac2f3f59c34320f0) after cherry-picked commit.

mburkart committed:
Enable readout of trigger efficiencies based on hadronic tau leptons identified with the DeepTau algorrithm.

mburkart committed:
Add input root files for efficiencies based on hadronic tau leptons identified with the DeepTau ID.

mburkart committed:
Add latest input root files for embedded.

mburkart committed:
Rename input root files for embedded SFs.

mburkart committed:
Update variable names of embedded related measurement.

mburkart committed:
Add test code for format of embedded SFs.

mburkart committed:
Adapt python code to read out embedded SFs.

mburkart committed:
Adapt C++ readout interface.

Change to hard coded file path as in the case of the python code. This
removes the need to give the file path as argument to the constructor
class. And allows for switch between embedded and MC scale factors by
boolean flag. The specific methods for embedded scale factors are now
obsolete and are removed.

mburkart committed:
Remove MC inputs from embedded input files.

mburkart committed:
Fix assertions in python interface.

Added correct assertion for working points for the different IDs. Fixed
bug in assertion for correct decay modes.

Commit:
Update README.md

Add description for scale factors for embedded samples.

kandrosov commented on src/TauTriggerSFs2017.cc:
```diff
@@ -56,38 +57,72 @@ int dmCheck(int dm )
+      assert(0);
+  }
+
+  std::string inputFileName = (boost::format("%s/src/TauAnalysisTools/TauTriggerSFs/data/tauTriggerEfficiencies%s.root") % std::getenv("CMSSW_BASE") % year).str();
```
- Input file names must be provided in the constructor. Not all analysis frameworks are CMSSW-based, so CMSSW_BASE variable could be not set.
- Like this it should also be backwards compatible, right? (`emb_sfs = false` by default)

mburkart replied:
* Does the first comment hold for the python interface as well? There it is implemented in the same way. If yes we would need to change this as well, right?
* Yes, the remark concerning the backwards compatibility only referred to the removed input file name as parameter of the constructor.

kandrosov replied:
Python interface can be used, for example, together with nanoAOD tools, which can be run standalone (without CMSSW). So, yes, better to change it as well. But you can keep back-compatibility: add file_name as an argument with default value None, and in case if it is not specified - load the default file name using CMSSW_BASE as it is currently done.

mburkart replied:
I made the changes to the python interface as you proposed. For the C++ interface I overloaded the constructor with a version allows to construct the readout class without specifying the input file name and added back the previous constructor of the readout class.

kandrosov commented on BuildFile.xml:
```diff
@@ -1,2 +1,3 @@
 <use name="root"/>
+<use name="boost"/>
```
If possible, I would try to avoid additional dependencies.

mburkart replied:
Done.

kandrosov commented on test/testTauTriggerSFs.cc:
```diff
@@ -0,0 +1,43 @@
+        {
+            for (std::string wp: working_points)
+            {
+                TauTriggerSFs2017 sf_reader = TauTriggerSFs2017(chan, year, wp, "DeepTau", true);
```
For completeness, I suggest testing the reader for both embedded and MC SFs. Same applies for `testTauTriggerSFs.py`.

mburkart replied:
Suggestion is implemented now.

mburkart committed:
Remove external dependency on boost::format.

mburkart committed:
Add readout tests for MC scale factors.

Additionally add readout of scale factor variations to tests.

mburkart committed:
Add file name as parameter of init in python.

Add an optional parameter to init method of readout class to give the
user the option to select an input file if no CMSSW release is used.

mburkart committed:
Add possibility to specify file name back in C++ interface.

Introduce a second constructor to C++ interface to allow for
construction of the readout class with optional specified file name
while keeping backward compatibility.
